### ISSUE TITLE
Promise.bind hard to use when thisArg may or may not be a thenable

### ISSUE THREAD

**Issue by reidrankin** [petkaantonov/bluebird#973]
This is less of a bug and more of a design issue. I've been debugging a bit of code that uses Sequelize, which uses bluebird internally, for two days now. I've found the issue, but placing blame is a bit harder.

Sequelize constructs "Instance" objects that represent database table rows and have methods like save() and reload() and such on them. It allows you to extend these objects with custom fields and methods. I'm dealing with an application that needs to wait until another program signals completion of a task by updating a database row, so I turned a Sequelize "Instance" into a thenable by defining a simple method that reloads the row every 5 seconds and waits for the appropriate signal to fire the callbacks. (My application also uses bluebird, and internally it uses a promise to dispatch the callbacks, but that's not relevant here.)

The problem comes because Sequelize expects `Promise.bind(this)` to explicitly bind to `this`. However, by turning the object into a thenable, I cause bluebird to change its behavior and attempt to bind to my thenable's result.

This is exactly what `Promise.bind()` says it will do on the tin, but I can't help but think that a single function that encapsulates two distinctly different behaviors and switches between them implicitly is a poor design.

Of course, any remediation here would be API-breaking, making it hard to know what to do. The options I've thought of include providing a `Promise.bindExplicit` version that does not change its behavior when presented with a thenable, or changing the behavior of the exisiting function so that it only binds-to-result when presented with a bluebird promise, and not simply a generic thenable (So you could do `Promise.bind(Promise.resolve(thenable))` to get the current behavior. Both options have their downsides.

Thoughts?

**Comment by petkaantonov** [comment]
Yes I hate that bind takes a promise a lot as well. It has been excruciatingly painful to implement for very dubious benefit, and it causes scenarios that don't make sense, for instance if the promise rejects then all `.finally` and `.catch` handlers might not have the expected `this` value and it just doesn't make sense. 

However, supporting promise but not thenable makes even less sense. Very often thenable is just a different instance of bluebird. And all apis always take a thenable wherever they take a promise.

**Comment by benjamingr** [comment]
I tend to also agree, I've seen a few uses with `.bind(thenable)` but I've never seen one where it felt like very useful design.
